cisagov/ScubaGear
# Add high risk application/service principal permissions into ScubaResults.json

## 🗣 Description ##




Adds a new module AADRiskyPermissionsHelper.psm1 which has the following functions: 
- Get-RiskyPermissionsJson
- Format-RiskyPermission
- Format-Credentials
- Merge-Credentials
- Get-ApplicationsWithRiskyPermissions
- Get-ServicePrincipalsWithRiskyPermissions
- Format-RiskyApplications
- Format-RiskyThirdPartyServicePrincipals

These functions serve the purpose of parsing application and service principal objects for risky permissions. The result can be found in ScubaResults.json with the following keys: `risky_applications` and `risky_third_party_service_principals`. 

Applications refer to applications created within a local tenant, while third-party service principals refer to service principals in your tenant that are owned by an external tenant.

New Microsoft commandlets added:
- [Get-MgBetaApplication](https://learn.microsoft.com/en-us/powershell/module/microsoft.graph.beta.applications/get-mgbetaapplication?view=graph-powershell-beta)
- [Get-MgBetaApplicationFederatedIdentityCredential](https://learn.microsoft.com/en-us/powershell/module/Microsoft.Graph.Beta.Applications/Get-MgBetaApplicationFederatedIdentityCredential?view=graph-powershell-beta)
- [Get-MgBetaServicePrincipal](https://learn.microsoft.com/en-us/powershell/module/microsoft.graph.beta.applications/get-mgbetaserviceprincipal?view=graph-powershell-beta)
- [Get-MgBetaServicePrincipalAppRoleAssignment](https://learn.microsoft.com/en-us/powershell/module/Microsoft.Graph.Beta.Applications/Get-MgBetaServicePrincipalAppRoleAssignment?view=graph-powershell-beta) (removed and updated to use batch endpoint, "/servicePrincipals/$ServicePrincipalId/appRoleAssignments")

## 💭 Motivation and context ##






Part of ongoing work for epic #1073

Closes #1327
Closes #1397 

Created #1463 as a follow-up issue

## 🧪 Testing ##





### How to run Pester tests

Checkout this PR's feature branch locally, then from the the root directory run the following commands:
```
Invoke-Pester -Output 'Detailed' -Path '.\PowerShell\ScubaGear\Testing\Unit\PowerShell\Providers\AADProvider\AADRiskyPermissionsHelper\'
```

Export-AADProvider.Tests.ps1 was modified since we call the above functions in Export-AADProvider.psm1. Run all AAD provider tests to ensure they pass:
```
Invoke-Pester -Output 'Detailed' -Path '.\PowerShell\ScubaGear\Testing\Unit\PowerShell\Providers\AADProvider\'
```

Create a new report, verify `risky_applications` and `risky_third_party_service_principals` are added in ScubaResults.json.

## ✅ Pre-approval checklist ##







## ✅ Pre-merge checklist ##






  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
    (Note: Only required for issues of size `Medium` or larger)

## ✅ Post-merge checklist ##
